patches/api/0387-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,286 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
+Date: Sun, 26 Jun 2022 20:24:16 +0200
+Subject: [PATCH] Add damage source wrapper
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/entity/damage/DamageAttribute.java b/src/main/java/com/destroystokyo/paper/entity/damage/DamageAttribute.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..e179dbf60ee5e3d9a2b079c1391778369f524675
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/entity/damage/DamageAttribute.java
+@@ -0,0 +1,38 @@
++package com.destroystokyo.paper.entity.damage;
++
++/**
++ * Damage attribute is useful to define the true nature of a damage
++ * If you don't know which one choose just take the generic one
++ * For some nature like the fire, this can directly change the global
++ * behavior of how the damage will be dealt.
++ * You cannot define multiple attribute on the same damage source
++ * Don't use damage attribute to reproduce vanilla damage
++ * use directly the shortcuts methods available in {{@link DamageOrigin}}
++ */
++public enum DamageAttribute {
++    SWEEP,
++    PROJECTILE,
++    // todo WIP
++    /**
++     * The damager who attack with an Explosion damage doesn't take the thorns damage from guardian
++     * Also the target receive no knockback
++     */
++    EXPLOSION,
++    /**
++     * Apply a fire damage, the target will burn and take
++     * fire damage if the game-rule {@link org.bukkit.GameRule#FIRE_DAMAGE} is enabled
++     * and that the player hasn't drunk a fire resistance potion.
++     * An armor with the fire protection enchantment can reduce the
++     * damage made.
++     *
++     * Some entities can have a special reaction to this damage:
++     * - {@link org.bukkit.entity.Item} can die immediately
++     * - {@link org.bukkit.entity.minecart.ExplosiveMinecart} can explode immediately
++     *
++     * Some entities (from nether mostly) can have developed an immunity to the fire
++     */
++    FIRE,
++    MAGIC,
++    FALL,
++    GENERIC
```

Comment:
Is it worth specifying these damage types as their own totally separate attributes? Vanilla supports these just fine, as they are just simple booleans.

I feel like especially in the case that multiple "attributes" can be used, it might be worth treating these independently as they currently are in vanilla.